statemachine.go

```diff
@@ -295,22 +296,25 @@ func (sm *StateMachine) setState(ctx context.Context, state State) error {
 	return sm.stateMutator(ctx, state)
 }
 
-func (sm *StateMachine) currentState(ctx context.Context) (sr *stateRepresentation, err error) {
-	var state State
-	state, err = sm.State(ctx)
-	if err == nil {
-		sr = sm.stateRepresentation(state)
+func (sm *StateMachine) currentState(ctx context.Context) (*stateRepresentation, error) {
+	state, err := sm.State(ctx)
+	if err != nil {
+		return nil, err
 	}
-	return
+	return sm.stateRepresentation(state), nil
 }
 
-func (sm *StateMachine) stateRepresentation(state State) (sr *stateRepresentation) {
-	var ok bool
-	if sr, ok = sm.stateConfig[state]; !ok {
+func (sm *StateMachine) stateRepresentation(state State) *stateRepresentation {
+	sm.stateMutex.RLock()
+	sr, ok := sm.stateConfig[state]
+	sm.stateMutex.RUnlock()
+	if !ok {
 		sr = newstateRepresentation(state)
+		sm.stateMutex.Lock()
 		sm.stateConfig[state] = sr
```

Comment:
should check whether state representation exists again.
like `double checked locking`

Reply:
Good catch, done!